File: service-product/src/main/kotlin/com/ultimate/ecommerce/serviceproduct/search/model/Product.kt
package com.ultimate.ecommerce.serviceproduct.search.model

import org.springframework.data.annotation.Id
import org.springframework.data.elasticsearch.annotations.Document

@Document(indexName = "product", createIndex = false)
data class Product(
    @Id
    val usin: String,
    val title: String,
    val description: String,
    val attributes: Map<String, String>,
    val images: List<String>,
    val ratings: List<Rating>,
    val sellOptions: List<SellOption>,
    val tag: String
)

data class Rating(
    val rate: Int,
    val amount: Int
)

data class SellOption(
    val price: Long,
    val currency: String,
    val type: String
)

File: service-ui/src/utils/product.ts
import { Product } from "../generated/service-product";

const MAXIMUM_RATING = 5;

export const countProductRating = (product: Product) => {
  const totalScore = product.ratings
    .map((rating) => rating.rate * rating.amount)
    .reduce((a, b) => a + b, 0);
  const maximumTotalScore = product.ratings
    .map((rating) => rating.amount * MAXIMUM_RATING)
    .reduce((a, b) => a + b, 0);
  if (maximumTotalScore === 0) {
    return 0;
  }
  return (totalScore / maximumTotalScore) * MAXIMUM_RATING;
};
